PurchasesCoreSwift/Purchasing/PurchasesOrchestrator.swift

```diff
@@ -0,0 +1,506 @@
+                      storeKitWrapper: StoreKitWrapper,
+                      systemInfo: SystemInfo,
+                      subscriberAttributesManager: SubscriberAttributesManager,
+                      operationDispatcher: OperationDispatcher,
+                      receiptFetcher: ReceiptFetcher,
+                      purchaserInfoManager: PurchaserInfoManager,
+                      backend: Backend,
+                      identityManager: IdentityManager,
+                      receiptParser: ReceiptParser,
+                      deviceCache: DeviceCache) {
+        self.productsManager = productsManager
+        self.storeKitWrapper = storeKitWrapper
+        self.systemInfo = systemInfo
+        self.subscriberAttributesManager = subscriberAttributesManager
+        self.operationDispatcher = operationDispatcher
+        self.receiptFetcher = receiptFetcher
+        self.purchaserInfoManager = purchaserInfoManager
+        self.backend = backend
+        self.identityManager = identityManager
+        self.receiptParser = receiptParser
+        self.deviceCache = deviceCache
+    }
+
+    @objc public func restoreTransactions(completion maybeCompletion: ((PurchaserInfo?, Error?) -> Void)?) {
+        syncPurchases(receiptRefreshPolicy: .always, isRestore: true, maybeCompletion: maybeCompletion)
+    }
+
+    @objc public func syncPurchases(completion maybeCompletion: ((PurchaserInfo?, Error?) -> Void)?) {
+        syncPurchases(receiptRefreshPolicy: .never,
+                      isRestore: allowSharingAppStoreAccount,
+                      maybeCompletion: maybeCompletion)
+    }
+
+    @objc public func products(withIdentifiers identifiers: [String],
+                               completion: @escaping ([SKProduct]) -> Void) {
+        let productIdentifiersSet = Set(identifiers)
+        guard !productIdentifiersSet.isEmpty else {
+            operationDispatcher.dispatchOnMainThread { completion([]) }
+            return
+        }
+
+        productsManager.products(withIdentifiers: productIdentifiersSet) { products in
+            self.operationDispatcher.dispatchOnMainThread {
+                completion(Array(products))
+            }
+        }
+    }
+
+    @available(iOS 12.2, macOS 10.14.4, watchOS 6.2, macCatalyst 13.0, tvOS 12.2, *)
+    @objc public func paymentDiscount(forProductDiscount productDiscount: SKProductDiscount,
+                                      product: SKProduct,
+                                      completion: @escaping (SKPaymentDiscount?, Error?) -> Void) {
+        receiptFetcher.receiptData(refreshPolicy: .onlyIfEmpty) { maybeReceiptData in
+            guard let receiptData = maybeReceiptData,
+                  !receiptData.isEmpty else {
+                      completion(nil, ErrorUtils.missingReceiptFileError())
+                      return
+                  }
+            guard let discountIdentifier = productDiscount.identifier else {
```

Comment:
could this be pulled before calling receiptFetcher.receiptData?

Reply:
Oh wow, yeah, that's smart.